Patch `cheby_equi_ripple_filter` for SciPy v1.16.0

gmgunter opened the pull request:
Closes #47

SciPy 1.16.0 was released recently with [this update](https://github.com/scipy/scipy/commit/f45ba7bd311ee60c5c2f74c272cd3cd6c1dece86) that seems to break the `scipy.signal.remez()` function when the `weight` argument is not `None`. This seems to be causing CI failures for us.

Let's exclude this SciPy version from the environment spec.

gmgunter commented:
I opened an upstream PR with a fix here: https://github.com/scipy/scipy/pull/23266

rad-eng-59 reviewed (approved):
LGTM. Thanks @gmgunter for the quick fix.
@gshiroma, The scipy folks has introduced a bug in line `836`. The problem is not on our side given the doc clearly says the optional input argument `weight` can be an array-like object. 
Alternatively, we can modify the `weight` to be a non-numpy container such as a python list rather than a numpy array on our side to avoid this scipy bug while still using the latest version of scipy. 
Let me know if you want to go with latter and I will open a PR to fix it.

rad-eng-59 commented:
@gmgunter Alternatively, as stated earlier, you can simply remove `np.array` from [this line](https://github.com/isce-framework/isce3/blob/968e78fb90e09fd69d34770fc6ab238ac2b0fdcb/python/packages/isce3/signal/fir_filter_func.py#L82) to let `scipy 1.16.0` go through successfully. That is
```python
...
weight=[1, weight_fact],
...
```
No need to limit scipy version in isce3.

Perhaps, we should remove `np.array()` from [*desired*, the third argument to remez](https://github.com/isce-framework/isce3/blob/968e78fb90e09fd69d34770fc6ab238ac2b0fdcb/python/packages/isce3/signal/fir_filter_func.py#L81) as well to avoid similar issue in the future.

gmgunter committed:
Work around a bug in `scipy.signal.remez` in SciPy v1.16.0

Pass the `weight` array as a list instead of a NumPy array to prevent
the `if weight` check in `scipy.signal.remez()` from raising a spurious
exception.

gmgunter commented:
Thanks, @rad-eng-59. I modified the PR to adopt your suggestion instead of excluding SciPy v1.16.0 from the environment spec.

rad-eng-59 reviewed (approved):
LGTM. Thank you so much @gmgunter and sorry for the trouble. 
I believe this is the second time the `remez` function being modified in SciPy :(

gshiroma reviewed (approved):
Thank you, @gmgunter !

gshiroma commented:
> LGTM. Thanks @gmgunter for the quick fix. @gshiroma, The scipy folks has introduced a bug in line `836`. The problem is not on our side given the doc clearly says the optional input argument `weight` can be an array-like object. Alternatively, we can modify the `weight` to be a non-numpy container such as a python list rather than a numpy array on our side to avoid this scipy bug while still using the latest version of scipy. Let me know if you want to go with latter and I will open a PR to fix it.

Sounds good. Thank you, @rad-eng-59 !

hfattahi reviewed (approved):
LGTM. Thanks @gmgunter.

gshiroma commented:
Uhm. Tests are still failing =/

gmgunter commented:
Looks like there's another bug in SciPy v1.16.0. The unit tests are now failing because the frequencies `f` computed by `scipy.signal.freqz` here:

https://github.com/isce-framework/isce3/blob/968e78fb90e09fd69d34770fc6ab238ac2b0fdcb/tests/python/packages/isce3/signal/fir_filter_func.py#L247

are coming out complex-valued now instead of real-valued.

Here's a snippet from the CI log:

```
x = array([0.36458333+0.j, 0.36653646+0.j, 0.36848958+0.j, 0.37044271+0.j,
       0.37239583+0.j, 0.37434896+0.j, 0.376302...1.35091146+0.j, 1.35286458+0.j, 1.35481771+0.j,
       1.35677083+0.j, 1.35872396+0.j, 1.36067708+0.j, 1.36263021+0.j])
fs = 1.0

    def wrapped_distance(x, fs=1.0):
>       xm = x % fs
             ^^^^^^
E       TypeError: ufunc 'remainder' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

https://github.com/isce-framework/isce3/actions/runs/16004235062/job/45146789710?pr=49#step:11:558

We're getting a `TypeError` here because the `%` operator doesn't support complex values.

I'm not sure if there's a similarly simple bandaid we can apply here? Otherwise, I can revert back to excluding SciPy v1.16.0 from the environment spec.

rad-eng-59 commented:
@gmgunter, please give me an hour or so to see if I can fix it and test it. Thank you

rad-eng-59 commented:
It is odd that the frequency values are complex type with zero imaginary part. The frequency response shall be complex given it is a single-side band-pass filter and thus whole frequency range is required to cover the full filter response (given asymmetric filter response within +/- Nyquist `= fs / 2`). However, the SciPy code also sets the datatype of frequency values, `f within [0, 1) ` to complex. I would say it is safe to simply use real part of frequency bins to avoid this bug that is 
```python
f, H = sig.freqz(h, fs=1.0, whole=True)
f = f.real
```
I have just tested it with SciPy `v1.15.2`  and `v1.16.0`. The difference is in `f` data type. Basically the real part in new version  is the same as older version with extra zero-filled imaginary part which I have found it very very strange! So, let's simply use the real part which works for both old and new versions.

Btw, in other instances like [this one](https://github.com/isce-framework/isce3/blob/968e78fb90e09fd69d34770fc6ab238ac2b0fdcb/tests/python/packages/isce3/signal/fir_filter_func.py#L215), the code uses absolute value for the frequency bins provided by `signal.freqz` just to be cautious , I guess.

what do you say?

gmgunter committed:
Add a patch for complex-valued frequencies from `scipy.signal.freqz`

This seems to be a bug introduced in SciPy v1.16.0

gmgunter commented:
Thanks @rad-eng-59 for investigating this. I've implemented the suggested patch and tested it locally with SciPy v1.13.1 and v1.16.0.

rad-eng-59 reviewed (approved):
LGTM. Hopefully this is the last change in `fir_filter_func.py`, at least for a while. 
Thank you @gmgunter for the fixes and apologies for all the trouble.
Thank you @gshiroma for reporting the issue in the first place and for your patience.

gmgunter commented:
Thanks @rad-eng-59 for all your help resolving this!
